File: csharp/src/Int2Dds/Core/Sample.cs
namespace Int2Dds.Core
{
    /// <summary>
    /// A data sample received from a DataReader.
    /// </summary>
    /// <typeparam name="T">The DDS data type.</typeparam>
    public readonly struct Sample<T> where T : class
    {
        /// <summary>The deserialized data, or null if not valid data (e.g., dispose/unregister notification).</summary>
        public T? Data { get; }

        /// <summary>True if this sample contains valid data; false for lifecycle notifications.</summary>
        public bool ValidData { get; }

        public Sample(T? data, bool validData)
        {
            Data = data;
            ValidData = validData;
        }
    }
}

File: java/api/src/main/java/com/intellectus/int2dds/core/Sample.java
package com.intellectus.int2dds.core;

import com.intellectus.int2dds.types.IDdsType;

/**
 * One received sample: the deserialized {@code data}, or {@code null} for a
 * sample whose {@link SampleInfo#validData()} is false (dispose/unregister),
 * paired with its {@link SampleInfo}.
 *
 * @param <T> the DDS data type.
 */
public final class Sample<T extends IDdsType> {

    private final T data;
    private final SampleInfo info;

    Sample(T data, SampleInfo info) {
        this.data = data;
        this.info = info;
    }

    /** The deserialized sample, or {@code null} when {@code info().validData()} is false. */
    public T data() {
        return data;
    }

    /** This sample's metadata; valid even when {@link #data()} is null. */
    public SampleInfo info() {
        return info;
    }
}
